src/StmtToViz.cpp

```diff
@@ -0,0 +1,2892 @@
+        for (auto const &entry : data_cost_inclusive) {
+            max_data_cost_inclusive = std::max(entry.second, max_data_cost_inclusive);
+        }
+    }
+
+    // Returns the compute cost of a node (estimated using simple op-counting)
+    int get_compute_cost(const IRNode *node, bool include_subtree_cost) {
+        internal_assert(node != nullptr) << "IRCostModel::get_compute_cost(): node is nullptr\n";
+
+        int cost = -1;
+        if (compute_cost.count(node)) {
+            cost = include_subtree_cost ? compute_cost_inclusive[node] : compute_cost[node];
+        } else {
+            internal_error << "IRCostModel::get_compute_cost(): cost lookup failed\n";
+        }
+
+        internal_assert(cost >= 0) << "Cost must not be negative.\n";
+        return cost;
+    }
+
+    // Returns the data movement cost of a node (the number of bits moved in load/store/shuffle ops)
+    int get_data_movement_cost(const IRNode *node, bool include_subtree_cost) {
+        internal_assert(node != nullptr) << "IRCostModel::get_data_movement_cost(): node is nullptr\n";
+
+        int cost = -1;
+        if (compute_cost.count(node)) {
+            cost = include_subtree_cost ? data_cost_inclusive[node] : data_cost[node];
+        } else {
+            internal_error << "IRCostModel::get_data_movement_cost(): cost lookup failed\n";
+        }
+
+        internal_assert(cost >= 0) << "Cost cost must not be negative.\n";
+        return cost;
+    }
+
+    // Returns the max compute cost of any node in the program
+    int get_max_compute_cost(bool include_subtree_cost) {
+        return include_subtree_cost ? max_compute_cost_inclusive : max_compute_cost;
+    }
+
+    // Returns the max data movement cost of any node in the program
+    int get_max_data_movement_cost(bool include_subtree_cost) {
+        return include_subtree_cost ? max_data_cost_inclusive : max_data_cost;
+    }
+
+private:
+    // Cost database. We track two costs:
+    //  - The line cost of a node is the sum of the node cost
+    //    plus the cost of any children that are printed on
+    //    a single line (since we display cost by each line in
+    //    the program)
+    //  - The inclusive cost is the cost of the entire sub-tree.
+    //    We display this cost when the user collapses a program
+    //    block in the IR.
+    std::map<const IRNode *, int> compute_cost;
+    std::map<const IRNode *, int> data_cost;
+
+    std::map<const IRNode *, int> compute_cost_inclusive;
+    std::map<const IRNode *, int> data_cost_inclusive;
+
+    // We also track the max costs to determine the cost color
+    // intensity for a given line of code
+    int max_compute_cost;
+    int max_data_cost;
+
+    int max_compute_cost_inclusive;
+    int max_data_cost_inclusive;
+
+    /* Utility functions to store node costs in the cost database */
+    void set_compute_costs(const IRNode *node, int node_cost, const std::vector<const IRNode *> &child_nodes) {
```

Comment:
I wonder if these should accepted initializer_lists instead of vectors for efficiency. They always seem to be brace-enclosed literals

Reply:
Ah, except for the call visitor

Reply:
> I wonder if these should accept initializer_lists instead of vectors for efficiency. They always seem to be brace-enclosed literals

<threadjack> when we eventually migrate to C++20 we should probably use `span` for all stuff like this

Reply:
I opened an issue so I remember to do it